File: src/kernel/src/platform/kalloc.c
#include <string.h>
#include <strings.h>
#include "kalloc.h"
#include "kprintf.h"

#define EfiMaxMemoryType      0x0000000e
#define EfiConventionalMemory 0x00000007

uint8_t *head = NULL;

void kalloc_debug()
{
}

void kalloc_init(kernel_memory_map_t *memory_map)
{
    kernel_memory_map_descriptor_t *d = memory_map->memory_map;

    uint64_t *start = NULL;
    uint64_t page_count = 0;

    for (uint64_t i = 0; i < memory_map->memory_map_size / memory_map->memory_map_descriptor_size; i++)
    {
        if (d->type == EfiConventionalMemory)
        {
            if (d->page_count > page_count)
            {
                start = (uint64_t *)d->physical_start;
                page_count = d->page_count;
            }
        }

        uint8_t *ptr = ((uint8_t *)d);
        ptr += memory_map->memory_map_descriptor_size;
        d = (kernel_memory_map_descriptor_t *)ptr;
    }

    kprintf(
        "kalloc_init: mapped %lu pages from 0x%lx to 0x%lx\n", 
        page_count,
        start,
        (uint64_t *)(((uint8_t *)start) + page_count * 4096));

    head = start;
}

void *kmalloc(size_t size)
{
    void *ptr = head;
    head += size;
    //kprintf("kmalloc: allocated %lu bytes at 0x%lx\n", size, ptr);
    return ptr;
}

void kfree(void *ptr)
{
    //kprintf("kmalloc: freeing ??? bytes at 0x%lx\n", ptr);   
}

void *kcalloc(size_t number, size_t size)
{
    void *new = head;
    head += number * size;
    bzero(new, number * size);
    //kprintf("kcalloc: allocated %lu bytes at 0x%lx\n", size, new);
    return new;
}

void *krealloc(void *ptr, size_t size)
{
    void *new = head;
    head += size;
    memcpy(new, ptr, size);
    //kprintf("krealloc: allocated %lu bytes at 0x%lx\n", size, new);
    return new;
}

File: src/kernel/src/kalloc.nim
import
  base,
  boot

const
  EfiMaxMemoryType* = 0x0000000e
  EfiConventionalMemory* = 0x00000007

var head: pointer

proc kallocInit* (mm: var BootMemoryMap) =
  printf("MEMORY MAP\n")
  printf("========================================\n")
  printf("map:                %p\n", cast[uint64](mm.map))
  printf("size:               %U\n", mm.size)
  printf("descriptorSize:     %U\n", mm.descriptorSize)
  printf("descriptorVersion:  %u\n", mm.descriptorVersion)
  printf("----------------------------------------\n")

  var d: ptr MemoryMapDescriptor = cast[ptr MemoryMapDescriptor](mm.map)
  var start: pointer = nil
  var pageCount: uint64 = 0

  for i in 0..mm.size div mm.descriptorSize:
    d = cast[ptr MemoryMapDescriptor](cast[pointer](cast[int](mm.map) + 
      cast[int](i * mm.descriptorSize)))

    if d.descriptorType == EfiConventionalMemory:
      printf("descriptorType: %u\n", d.descriptorType)
      printf("physicalStart:  %p\n", cast[uint64](d.physicalStart))
      printf("virtualStart:   %p\n", cast[uint64](d.virtualStart))
      printf("pageCount:      %U\n", d.pageCount)
      printf("attributes:     %U\n", d.attributes)
      printf("----------------------------------------\n")

      if pageCount < d.pageCount:
        start = cast[pointer](d.physicalStart)
        pageCount = d.pageCount

  head = start

  printf("kallocInit: mapped %U pages at %p\n", pageCount, cast[uint64](start))

proc malloc(size: uint64): pointer {.exportc.} =
  let p: pointer = head
  printf("malloc: mapping %U bytes at %p\n", size, cast[uint64](p))
  head = head + size
  return p

proc free(p: pointer) {.exportc.} =
  printf("free: %p\n", cast[uint64](p))
  discard

proc calloc(number: uint64, size: uint64): pointer {.exportc.} =
  let p: pointer = head
  printf("calloc: mapping %U bytes at %p\n", number * size, cast[uint64](p))
  head = head + (number * size)
  return p

proc realloc(p: pointer, size: uint64): pointer {.exportc.} =
  let n: pointer = head
  copyMem(n, p, size)
  printf("realloc: mapping %p to %U bytes at %p\n", cast[uint64](p), size, cast[uint64](n))
  head = head + size
  return n
